File: project1/board.rb
# Author: Ryan O'Donovan
# Represent the physical play space

class BoardMaker


  require_relative 'deck_maker.rb'
  require_relative 'card.rb'
  require_relative 'CardParser.rb'

attr_accessor :score 
attr_reader :board, :deck

  def initialize

    # Set up the board
  @board = Array.new 12
    
  @deck = DeckMaker.new

  @score = 0


  @deck.shuffle

  x = 0

    # Place 12 cards onto the board
  until x == 12

    @board[x] = deck.returnOne

      x += 1

  end
  
  end

  def display

  x = 0

  i = 1

  until x == @board.length

    # 3 cards per row
    if i < 3 

      print "%-15s" %[@board[x].return_card]

      x += 1

      i += 1

    else
      # Make new line after 3rd card has been placed and start again
      print "%-15s\n" %[@board[x].return_card]

      x += 1

      i = 1

    end

  end

  print "Score: %d   | Cards Left: %d\n" %[@score, @deck.cardsLeft]

end

  # Author : Samiul Islam
  # Check every possible combination of cards on the board to see if they make a set
def hasSet

  # [headIndex][nextIndex][latestIndex][][] --> [headIndex][nextIndex][][latestIndex][]--> [headIndex][nextIndex][][][latestIndex]
  # [headIndex][][nextIndex][latestIndex][] ... --> [][headIndex][nextIndex][latestIndex][]
  headIndex = 0
  nextIndex = 1
  latestIndex = 2

  # Move each head around the whole board to check all combinations
  until headIndex == board.length
    until nextIndex == board.length
      until latestIndex == board.length
        result =  CardParsing::setParser([board[headIndex], board[nextIndex], board[latestIndex]])
        if result
          return true
        end
        # Move the 3rd cards pointer down
        latestIndex += 1
      end

      # Move the 2nd cards pointer down, reset 3rd cards pointer to point ahead of 2nd cards
      nextIndex += 1
      latestIndex = (nextIndex+1)
    end

    #Move 1st cards pointer down, reset other pointers
    headIndex += 1
    nextIndex = (headIndex+1)
    latestIndex = (nextIndex+1)
  end

  return false
end

  # Add 3 more cards to the board from the deck
def replenish

 i = 0

 until i == 3

  newCard = @deck.returnOne

  @board.push newCard

  i += 1

 end

end


# Remove a subset of cards from the board
def remove(cards)

  i = 0

  @board = @board - cards
end
end

File: project2/board.js
/* Author: Ryan O'Donovan
 * Represents the physical play space */

	import './CardParser.js';

	import * as Deck from './deck-maker.js';

	class BoardMaker {


	/* New is boardMaker's constructor, convention to capitalize	*/


	function New() {

		this.board = new Array(12);
		deck = new Deck();

		this.score = 0;

		deck.shuffle();

		var x = 0;

		while (x < 12){
			
			this.board[x] = deck.returnOne(); 

			x += 1;
		}


	}

	

	function dispayBoard() {
		

		var i = 0;

		var length = this.board.length;

		var tr;


		while(i < length){

			tr += "<tr>";

			tr += "<td>" + this.board[i]+ "</td>";

			i += 1;

			tr += "<td>" + this.board[i]+ "</td>";

                        i += 1;

			tr += "<td>" + this.board[i]+ "</td>";

                        i += 1;

			tr += "</tr>";

		}

		tr += "</tr>";


		var table = document.getElementById("playing-field");

		table.innerHTML = tr;
		



	}




	


	function hasSet() {

		var headIndex = 0;
		var nextIndex = 1;
		var latestIndex = 2;

	/ Move each head around the board to check all combinations/

		while (headIndex < this.length) {
			while (nextIndex < this.length) {
				while (latestIndex < this.length) {

					var result = setParser(this.board[headIndex], 
						this.board[nextIndex], this.board[latestIndex]) 

					if (result) return true;


					latestIndex +=1;
			
				}

				nextIndex += 1;

				latestIndex = (nextIndex + 1);

			}

			headIndex += 1;
			nextIndex = (headIndex + 1);
			latestIndex = (nextIndex + 1);

		}

		return false;

	}

	function replenishCards() {

		var i = 0;

		while(i < 3) {

			var newCard = deck.returnOne;

			this.board.push(newCard);

			i += 1;

		}

	}


	function removeEntry(cards) {


	/* Since we have all unique cards, there is no need to have an if block
	 * to see if this card is in the array.					*/

		var card1 = this.board.indexOf(cards[0]);
		var card2 = this.board.indexOf(cards[1]);
		var card3 = this.board.indexOf(cards[2]);

		this.board.splice(card1, 1);
		this.board.splice(card2, 1);
		this.board.splice(card3, 1);

	}

}
